File: script/kubernetes/smoke/http_client.rb
# frozen_string_literal: true

require "json"
require "net/http"
require "uri"

module KubernetesSmoke
  class HttpClient
    Response = Struct.new(:status, :body, :headers)

    def initialize(timeout:)
      @timeout = timeout
    end

    def get(url, headers: {})
      request("GET", url, headers: headers)
    end

    def post(url, headers: {}, json: nil)
      request("POST", url, headers: headers, json: json)
    end

    def delete(url, headers: {})
      request("DELETE", url, headers: headers)
    end

    def request(method, url, headers: {}, json: nil)
      uri = URI.parse(url)
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = uri.scheme == "https"
      http.open_timeout = @timeout
      http.read_timeout = @timeout

      request = request_class(method).new(uri)
      headers.each { |key, value| request[key] = value }
      if json
        request["Content-Type"] = "application/json"
        request.body = JSON.generate(json)
      end

      response = http.request(request)
      Response.new(response.code.to_i, response.body.to_s, response.to_hash)
    rescue URI::InvalidURIError => error
      raise CheckFailure, "invalid URL #{url.inspect}: #{error.message}"
    rescue IOError, SystemCallError, Timeout::Error, SocketError => error
      raise CheckFailure, "HTTP request to #{url} failed: #{error.class}: #{error.message}"
    end

    private

    def request_class(method)
      case method
      when "GET"
        Net::HTTP::Get
      when "POST"
        Net::HTTP::Post
      when "DELETE"
        Net::HTTP::Delete
      else
        raise ArgumentError, "unsupported HTTP method #{method}"
      end
    end
  end
end

File: script/kubernetes/smoke-test/http_client.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"time"
)

type Response struct {
	Status  int
	Body    string
	Headers http.Header
}

type HTTPClient struct {
	client *http.Client
}

func NewHTTPClient(timeoutSeconds int) *HTTPClient {
	return &HTTPClient{
		client: &http.Client{Timeout: time.Duration(timeoutSeconds) * time.Second},
	}
}

func (client *HTTPClient) Get(targetURL string, headers map[string]string) (Response, error) {
	return client.request(http.MethodGet, targetURL, headers, nil)
}

func (client *HTTPClient) Post(targetURL string, headers map[string]string, jsonBody any) (Response, error) {
	return client.request(http.MethodPost, targetURL, headers, jsonBody)
}

func (client *HTTPClient) Delete(targetURL string, headers map[string]string) (Response, error) {
	return client.request(http.MethodDelete, targetURL, headers, nil)
}

func (client *HTTPClient) request(method string, targetURL string, headers map[string]string, jsonBody any) (Response, error) {
	parsed, err := url.Parse(targetURL)
	if err != nil {
		return Response{}, CheckFailure(fmt.Sprintf("invalid URL %q: %s", targetURL, err))
	}
	if parsed.Scheme == "" || parsed.Host == "" {
		return Response{}, CheckFailure(fmt.Sprintf("invalid URL %q: missing scheme or host", targetURL))
	}

	var body io.Reader
	if jsonBody != nil {
		payload, err := json.Marshal(jsonBody)
		if err != nil {
			return Response{}, err
		}
		body = bytes.NewReader(payload)
	}

	request, err := http.NewRequest(method, targetURL, body)
	if err != nil {
		return Response{}, CheckFailure(fmt.Sprintf("invalid URL %q: %s", targetURL, err))
	}
	for key, value := range headers {
		request.Header.Set(key, value)
	}
	if jsonBody != nil {
		request.Header.Set("Content-Type", "application/json")
	}

	response, err := client.client.Do(request)
	if err != nil {
		return Response{}, httpFailure(targetURL, err)
	}
	defer response.Body.Close()

	responseBody, err := io.ReadAll(response.Body)
	if err != nil {
		return Response{}, httpFailure(targetURL, err)
	}
	return Response{
		Status:  response.StatusCode,
		Body:    string(responseBody),
		Headers: response.Header,
	}, nil
}

func httpFailure(targetURL string, err error) error {
	if netErr, ok := err.(net.Error); ok {
		return CheckFailure(fmt.Sprintf("HTTP request to %s failed: %T: %s", targetURL, netErr, netErr.Error()))
	}
	return CheckFailure(fmt.Sprintf("HTTP request to %s failed: %T: %s", targetURL, err, err.Error()))
}
